File: problems/problem 3/test_cases.java
import java.util.*;

public class test_cases {
    private static int failures = 0;

    // normalize string: sort lines so order doesn't matter
    private static String normalize(String s) {
        s = s.trim();
        if (s.isEmpty()) return "";
        String[] lines = s.split("\\r?\\n");
        Arrays.sort(lines);
        return String.join("\n", lines) + "\n";
    }

    private static void check(int[] arr, int target, String expected) {
         String[] pairs = solution.twoSumPairs(arr, target);
        String output = String.join("\n", pairs) + "\n"; // convert array to string
        String normOut = normalize(output);
        String normExp = normalize(expected);

        if (normOut.equals(normExp)) {
            System.out.println("PASS");
        }   
        else {
            System.out.println("FAIL\nExpected:\n" + normExp + "Got:\n" + normOut);
            failures++;
        }
}


    public static void main(String[] args) {
        check(new int[]{1,2,3,2,4}, 4, "0 2\n1 3\n");
        check(new int[]{3,3,4,7,5,2}, 10, "1 3\n2 4\n");
        check(new int[]{0,0,0,0}, 0, "0 1\n0 2\n0 3\n1 2\n1 3\n2 3\n");
        check(new int[]{-1,-2,-3,-4,-5}, -8, "2 4\n");

        if (failures > 0) {
            System.exit(1); // non-zero = failure
        } else {
            System.exit(0); // zero = success
        }
    }
}

File: problems/problem 3/test_cases.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


char* twoSumPairs(int *arr, int n, int target);

int failures = 0;

void normalize(char *s) {

    char *lines[1000];
    int count=0;
    char *tok = strtok(s, "\n");
    while (tok) {
        lines[count++] = tok;
        tok = strtok(NULL, "\n");
    }
    for (int i=0; i<count; i++) {
        for (int j=i+1; j<count; j++) {
            if (strcmp(lines[i], lines[j])>0) {
                char *tmp = lines[i];
                lines[i] = lines[j];
                lines[j] = tmp;
            }
        }
    }
    s[0]='\0';
    for (int i=0; i<count; i++) {
        strcat(s, lines[i]);
        strcat(s, "\n");
    }
}

void check(int arr[], int n, int target, char *expected) {
    char *output = twoSumPairs(arr,n,target);
    normalize(output);
    normalize(expected);
    if (strcmp(output, expected)==0) {
        printf("PASS\n");
    } else {
        printf("FAIL Expected=\n%sGot=\n%s", expected, output);
        failures++;
    }
    free(output);
}

int main() {
    int arr1[] = {1,2,3,2,4};
    check(arr1, 5, 4, strdup("0 2\n1 3\n"));

    int arr2[] = {3,3,4,7,5,2};
    check(arr2, 6, 10, strdup("1 3\n2 4\n"));

    int arr3[] = {0,0,0,0};
    check(arr3, 4, 0, strdup("0 1\n0 2\n0 3\n1 2\n1 3\n2 3\n"));

    int arr4[] = {-1,-2,-3,-4,-5};
    check(arr4, 5, -8, strdup("2 4\n"));

    if (failures>0) return 1;
    return 0;
}
